### DISCUSSION TITLE
Practical 2.2 problem

### DISCUSSION THREAD

**Discussion by Rawizard** [IBAHCM/RPiR#175] [Archive] [answered]
Hello, 

When I try and run my step sis simulation the herd.df output only has one observation in it and i get this error: error in match.names(clabs, names(iv)).

Link to my practical 2 repo:

https://github.com/IBAHCM/jamiecrowther_series02 

Any help would be greatly appreciated!

**Comment by soniamitchell** [comment, marked as answer]
Hi @Rawizard, you've set up the `ecoli.transmission` vector perfectly, that great. But on line 35 you reference `ecoli.transmission` rather than `new.transmission`. This means that on line 35, `r0` is a vector rather than a single number. The same is true on line 53, where you're giving 4 rates rather than 1.

I'd also suggest removing `start.time`, `end.time`, `timesteps`, and `herd.df` from your loop (lines 39-47), as they don't rely on `new.transmission`. Instead, each time you run through the loop with a new `new.transmission`, you're re-writing these objects with the same contents. This isn't related to your error, but is good practice.

A good way to help identify bugs like this is to set `new.transmission = 1/2` in your console and then run through your code line by line (inside your `ecoli.transmission` loop) to help you identify your problem.

Once you get to your `timesteps` loop, you should find your error:

```
Error in match.names(clabs, names(xi)) : 
  names do not match previous names
```

At this point, `herd.df` should look like this: 

```
     susceptibles infecteds
1           98         2
```

.. and if you set `new.time = 1` in the console, then run through your `timesteps` loop line by line, hopefully `next.population` should look like this:

```
      susceptible infected
1    97.68667 2.313333
```

If you compare these two objects you'll see that the column names are different. On line 55, `rbind()` is upset because it doesn't want to stick rows together if their column names are different. If you fix this, your error will go away. 

But you have another problem because each time you go through your outer loop, `herd.df` will be overwritten. A hint to solve this is to either use a list, with a different `herd.df` in each element of that list... or to add an extra column identifying the transmission rate used and using `rbind` to stick each `herd.df` together after your tilmestep loop.

Really good effort though! Well done :)

**Comment by Rawizard** [comment]
Hi Sonia,

Thanks for getting back to me!

 I have fixed the problems you have identified apart from the final comment you have made about over writing herd.df in my outer loop. I am a bit confused by this?

I have added a transmission column to (not entirely sure thats what you meant??):

herd.df <- data.frame(susceptible = initial.susceptibles,
                      infected = initial.infecteds,
                      transmission = ecoli.transmission)

but i now get the error:

Error in `$<-.data.frame`(`*tmp*`, "time", value = c(0, 1, 2, 3, 4, 5,  :
  replacement has 101 rows, data has 4

Cheers,

Jamie


On 28 Jan 2021, at 13:25, Sonia Mitchell <notifications@github.com<mailto:notifications@github.com>> wrote:



Hi @Rawizard<https://github.com/Rawizard>, you've set up the ecoli.transmission vector perfectly, that great. But on line 35 you reference ecoli.transmission rather than new.transmission. This means that on line 35, r0 is a vector rather than a single number. The same is true on line 53, where you're giving 4 rates rather than 1.

I'd also suggest removing start.time, end.time, timesteps, and herd.df from your loop (lines 39-47), as they don't rely on new.transmission. Instead, each time you run through the loop with a new new.transmission, you're re-writing this object with the same contents.

A good way to help identify bugs like this is to set new.transmission = 1/2 in your console and then running your code line by line (inside your ecoli.transmission loop) to help you identify your problem.

Once you get to your timesteps loop, you should find your error:

Error in match.names(clabs, names(xi)) :
  names do not match previous names


At this point, herd.df should look like this:

     susceptibles infecteds
1           98         2


.. and if you set new.time = 1 in the console, then run through your timesteps loop line by line, hopefully next.population should look like this:

      susceptible infected
1    97.68667 2.313333


If you compare these two objects you'll see that the column names are different. On line 55, rbind() is upset because it doesn't want to stick rows together if their column names are different. If you fix this, your error will go away.

But you have another problem because each time you go through your outer loop, herd.df will be overwritten. A hint to solve this is to either use a list, with a different herd.df in each element of that list... or to add an extra column identifying the transmission rate used and using rbind to stick each herd.df together after your tilmestep loop.

**Reply by soniamitchell** [reply]
Hi @Rawizard, if you run through your code line by line, you should find that `next.population` on line 55 is empty. This is because you forgot to change your column names in `step_deterministic_SIS()`.

As for the next bit, the code is a little different now so you're no longer overwriting `herd.df`. But there's still a problem, which I'll explain... but first, please remove line 37, that's not what I meant.

You create your `herd.df` on line 35 and for each `new.transmission` and `new.time` you add a new row to it (on line 59). Before, when we only had the `timesteps` loop, we added a new column, `time` (on line 65) so that we could identify which row was being printed in each time step. I want you to do a similar thing for each ecoli transmission rate. Does that make sense?

**Comment by Rawizard** [comment]
Hi Sonia,

Sorry I have no fixed my step_deterministic_SIS().

I see what you mean and I have now done what I think is the equivalent for transmission but it doesn’t seem to work.

I am also confused as to why there is only one observation in my next.population?

Thank you,

Jamie

On 28 Jan 2021, at 15:24, Sonia Mitchell <notifications@github.com<mailto:notifications@github.com>> wrote:



Hi @Rawizard<https://github.com/Rawizard>, if you run through your code line by line, you should find that next.population on line 55 is empty. This is because you forgot to change your column names in step_deterministic_SIS().

As for the next bit, the code is a little different now so you're no longer overwriting herd.df. But there's still a problem, let me explain... but first, please remove line 37, that's not what I meant.

You create your herd.df on line 35 and for each new.transmission and new.time you add a new row to it (on line 59). Before, when we only had the timesteps loop, we added a new column, time (on line 65) so that we could identify which row was being printed in each time step. I want you to do a similar thing for each ecoli transmission rate. Does that make sense?

**Reply by soniamitchell** [reply]
Hi @Rawizard, there is only one observation in `next.population` because that's what `step_deterministic_SIS()` does. Actually, `next.population` should contain the "next population", or "the population at the next time step". It is `herd.df` you should be looking at. 

Try this:

1. Run line 25, then look at `herd.df` (this is time step 0)
2. Set `new.transmission = 1/2` in the console (we're calculating at this transmission rate)
3. Run lines 47-59, then look at `herd.df` (which now contains all of the observations for all timesteps for `new.transmission = 1/2`)
4. Now run line 64, and again look at `herd.df`. Now highlight only this bit `c(start.time, timesteps)` and run it.
5. Now replace line 65 with `herd.df$transmission <- new.transmission`
6. Run line 65 and look at `herd.df`

Does that make sense? Running through your code line by line and checking what happens to the objects is a great way to find problems in your code :)

And actually.. I made a mistake! put this bit back in the beginning of your `ecoli.transmission` loop:
```
  herd.df <- data.frame(susceptible = initial.susceptibles,
                        infected = initial.infecteds)
```
It actually does need to be reset for each transmission rate (because these are separate experiments)... which means you don't need a new $transission column. But it's nice to have for clarity, I think. Sorry~ Anyway, hopefully this will help your debugging in the future! :)

**Comment by Rawizard** [comment]
Hi Sonia,

Thank you!!

That now works and I see what you mean about running individual lines, thought I had been doing that but clearly I wasn’t.

Cheers,

Jamie


On 28 Jan 2021, at 16:56, Sonia Mitchell <notifications@github.com<mailto:notifications@github.com>> wrote:

new.transmission = 1/2

**Reply by soniamitchell** [reply]
That's great! Running through code line by line sounds simple, but after having to explain it in text, I can see that it's slightly more involved than that! If you get into the habit of doing it though, it should really help those debugging skills :)
